### PR TITLE
Externalize forward declarations for Track object

### PR DESCRIPTION
**Pull request by daschuer**
This finishes the work started by @uklotzde in: 
https://github.com/mixxxdj/mixxx/pull/3136

### CODE REVIEW

**Comment by uklotzde** [comment]
Thanks for continuing this. I hesitated from doing the full work before reaching consensus.

With the simple recommendation or rule to include only the corresponding "_decl.h" file in other header files our build dependencies could be reduced substantially for some central classes. Then all code that needs access the full ".h" file must be defined in the ".cpp" file, i.e. no inlining of code that depends on implementation details of those pervasive classes.

**Comment by daschuer** [comment]
Yes, but maybe we have a disagreement about pure forward declarations. Lets clarify: 

I think we must not require create decl.h files if a pure "class Xyz;" statement is sufficient.
I prefer this, when possible (no smart pointer is involved) because the purpose is unambiguous, compared to a #include statement that requires to look into an other file when in doubt.

**Comment by daschuer** [comment]
strange .. why it works sometimes: 
```
[ 95%] Building CXX object CMakeFiles/mixxx-test.dir/src/test/looping_control_test.cpp.o
/home/appveyor/projects/mixxx/src/test/librarytest.cpp: In function ‘void {anonymous}::deleteTrack(Track*)’:
/home/appveyor/projects/mixxx/src/test/librarytest.cpp:10:12: error: possible problem detected in invocation of delete operator: [-Werror=delete-incomplete]
     delete pTrack;
            ^~~~~~
/home/appveyor/projects/mixxx/src/test/librarytest.cpp:7:25: error: ‘pTrack’ has incomplete type [-Werror]
 void deleteTrack(Track* pTrack) {
                         ^~~~~~
In file included from /home/appveyor/projects/mixxx/src/library/dao/cuedao.h:6:0,
                 from /home/appveyor/projects/mixxx/src/library/trackcollection.h:10,
                 from /home/appveyor/projects/mixxx/src/test/librarytest.h:7,
                 from /home/appveyor/projects/mixxx/src/test/librarytest.cpp:1:
/home/appveyor/projects/mixxx/src/track/cue.h:15:7: note: forward declaration of ‘class Track’
 class Track;
       ^~~~~
/home/appveyor/projects/mixxx/src/test/librarytest.cpp:10:12: note: neither the destructor nor the class-specific operator delete will be called, even if they are declared when the class is defined
     delete pTrack;
            ^~~~~~
cc1plus: all warnings being treated as errors
```

File: src/track/track_decl.h [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+#include <QMetaType>
+#include <memory>
+
+class Track;
+
+typedef std::shared_ptr<Track> TrackPointer;
+typedef std::weak_ptr<Track> TrackWeakPointer;
+typedef QList<TrackPointer> TrackPointerList;
+
+enum class ExportTrackMetadataResult {
```
Comment (uklotzde) [inline src/track/track_decl.h:15]: Remark: I have noticed that this enum class does not really belong here. Refactoring the API such that the enumeration does not belong to the `Track` object is a future task. The extraction of all definitions into the new `track_decl.h` file just revealed this issue.

**Comment by uklotzde** [comment]
I would like to limit the usage of explicit forward declarations as much as possible. They become tedious when moving definitions consequently into nested namespaces that would need to be repeated for those forward declarations.

**Comment by uklotzde** [comment]
Ready after CI builds have passed.

**Comment by daschuer** [comment]
> I would like to limit the usage of explicit forward declarations as much as possible. They become tedious when moving definitions consequently into nested namespaces that would need to be repeated for those forward declarations.

This is just a one find and replace action more.  Remember that you edit the code one time but read it multiple time.

**Review by uklotzde — APPROVED** [review]

**Comment by uklotzde** [comment]
Thank you for finishing this tedious task! Let's see if this will help to improve a developer's life. If it proves to be useful we could apply this pattern to the other candidates.

LGTM

**Comment by uklotzde** [comment]
I'll take care of merging this into master, some conflicts.

**Pull request merged**